File: app-java/src/main/java/science/monke/internal/order/control/OrderIdGenerator.java
package science.monke.internal.order.control;

import lombok.SneakyThrows;
import lombok.extern.log4j.Log4j2;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import science.monke.spring.config.CustomProperties;
import science.monke.spring.config.Environment;
import science.monke.spring.config.Region;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.OffsetDateTime;
import java.util.Random;

@Log4j2
@Component
public class OrderIdGenerator {

  private static final Random RANDOM = new Random();

  private final CustomProperties customProperties;

  @Autowired
  public OrderIdGenerator(final CustomProperties customProperties) {
    this.customProperties = customProperties;
  }

  public String generateOrderId() {
    return generateOrderId(OffsetDateTime.now(), generateRandomString());
  }

  @SneakyThrows
  public String generateOrderId(final OffsetDateTime offsetDateTime, final String salt) {
    final MessageDigest digest = MessageDigest.getInstance("SHA3-256");
    final Region region = customProperties.getRegion();
    final Environment environment = customProperties.getEnvironment();
    final String valueToHash = region.name() + environment.name() + offsetDateTime + salt;
    final byte[] hashBytes = digest.digest((valueToHash).getBytes(StandardCharsets.UTF_8));
    final String sha3Hex = bytesToHex(hashBytes);
    final String orderId =
        sha3Hex.substring(0, 8)
            + "-"
            + region.name()
            + "-"
            + environment.name()
            + "-"
            + sha3Hex.substring(32, 40);
    log.info("Generated new order id: {}", orderId);
    return orderId;
  }

  public String generateRandomString() {
    final byte[] array = new byte[4];
    RANDOM.nextBytes(array);
    return bytesToHex(array);
  }

  private String bytesToHex(final byte[] hash) {
    final StringBuilder hexString = new StringBuilder(2 * hash.length);
    for (final byte b : hash) {
      final String hex = Integer.toHexString(0xff & b);
      if (hex.length() == 1) {
        hexString.append('0');
      }
      hexString.append(hex);
    }
    return hexString.toString();
  }
}

File: app-kotlin/src/main/kotlin/science/monke/internal/order/control/OrderIdGenerator.kt
package science.monke.internal.order.control

import org.springframework.stereotype.Component
import science.monke.spring.properties.Environment
import science.monke.spring.properties.Region
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.security.SecureRandom
import java.time.OffsetDateTime
import java.util.*

@Component
class OrderIdGenerator {
    fun random(region: Region, environment: Environment): String {
        return of(region, environment, OffsetDateTime.now(), randomSalt())
    }

    fun of(region: Region, environment: Environment, offsetDateTime: OffsetDateTime, salt: String): String {
        val bytes: ByteArray =
            (region.name + environment.name + offsetDateTime + salt).toByteArray(StandardCharsets.UTF_8)
        val uuid: UUID = UUID.nameUUIDFromBytes(bytes)
        val bas64UUID: String = uuidToBase64(uuid)

        var identifier = "-$region-$environment-"
        if (environment == Environment.PROD) {
            identifier = "-$region-"
        }

        val bas64UUIDLength: Int = bas64UUID.length / 2

        return bas64UUID.substring(0, bas64UUIDLength) + identifier + bas64UUID.substring(bas64UUIDLength)
    }

    private fun randomSalt(): String {
        val byteArray = ByteArray(4)
        SecureRandom.getInstanceStrong().nextBytes(byteArray)
        return byteArray.contentToString()
    }

    private fun uuidToBase64(uuid: UUID): String {
        val base64UUID = Base64.getUrlEncoder().encodeToString(uuid.toBytes())
        return base64UUID.replace('-', '!').replace('_', '*').substring(0, base64UUID.length - 2)
    }

    private fun UUID.toBytes(): ByteArray {
        val byteBuffer = ByteBuffer.wrap(ByteArray(16))
        byteBuffer.putLong(mostSignificantBits)
        byteBuffer.putLong(leastSignificantBits)
        return byteBuffer.array()
    }
}

